File: 120-triangle/triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int n=triangle.size();
        int c=INT_MAX;
        int m=triangle[n-1].size();
        vector<vector<int>> cost(n+1,vector<int>(m+1,INT_MAX));
        queue<pair<int,int>>q;
        q.push({0,0});
        cost[0][0]=triangle[0][0];
        while(!q.empty())
        {
            int x=q.front().first;
            int y=q.front().second;
            q.pop();
            
            if (x + 1 < n) { 
                if (y+1 < triangle[x+1].size() && 
                    cost[x+1][y+1] > cost[x][y] + triangle[x+1][y+1]) {
                    q.push({x+1,y+1});
                    cost[x+1][y+1] = cost[x][y] + triangle[x+1][y+1];
                }
                if (cost[x+1][y] > cost[x][y] + triangle[x+1][y]) {
                    q.push({x+1,y});
                    cost[x+1][y] = cost[x][y] + triangle[x+1][y];
                }
            }
        }
        for(int i=0;i<m;i++)
        {
            c=min(c,cost[n-1][i]);
        }
        return c;
    }
};

File: 120-triangle/triangle.java
class Solution {
    // public int helper(int[][] mat,int i,int j,int[][] dp)
    // {
    //     if(i==mat.length-1)
    //     {
    //         return mat[i][j];
    //     }
    //     if(j>=mat[i].length)
    //     {
    //         return Integer.MAX_VALUE;
    //     }
    //     if(dp[i][j]!=-1)
    //     {
    //         return dp[i][j];
    //     }
    //     int a= helper(mat,i+1,j+1,dp);
    //     int b=helper(mat,i+1,j,dp);

    //     return dp[i][j]= mat[i][j] + Math.min(a,b);
        
    // }
    public int minimumTotal(List<List<Integer>> triangle) {
        int n=triangle.size();
        int[][] dp = new int[n][n];
        for(int i=0;i<n;i++)
        {
            Arrays.fill(dp[i],-1);
        }
        dp[0][0]=triangle.get(0).get(0);
        for(int i=1;i<n;i++)
        {
            for(int j=0;j<=i;j++)
            {
                int a=Integer.MAX_VALUE;
                int b=Integer.MAX_VALUE;
                if(j<i) a= dp[i-1][j];
                if(j-1>=0)
                {
                    b=dp[i-1][j-1];
                }
                dp[i][j]=triangle.get(i).get(j) + Math.min(a,b);
            }
        }
        int x=Integer.MAX_VALUE;
        for(int i : dp[n-1])
        {
            x=Math.min(x,i);
        }
        return x;
    }
}